homeassistant/components/lgthinq/switch.py

```diff
@@ -0,0 +1,69 @@
+"""Support for switch entities."""
+
+from __future__ import annotations
+
+from collections.abc import Collection
+import logging
+from typing import Any
+
+from homeassistant.components.switch import SwitchDeviceClass, SwitchEntity
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import ThinqConfigEntry
+from .const import THINQ_DEVICE_ADDED
+from .device import LGDevice
+from .entity_helpers import ThinQEntity, ThinQSwitchEntityDescription
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up an entry for switch platform."""
+
+    @callback
+    def async_add_devices(devices: Collection[LGDevice]) -> None:
+        """Add switch entities."""
+        async_add_entities(ThinQSwitchEntity.create_entities(devices))
+
+    async_add_devices(entry.runtime_data.device_map.values())
+
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, THINQ_DEVICE_ADDED, async_add_devices)
+    )
```

Comment:
I think the integration could be way more maintainable if it was using the common architecture of an integration.

`async_setup_entry` in `__init__.py` -> Everything needed to setup the integration
`async_setup_entry` in the platforms -> Everything needed to add entities and to add them on runtime
The entity classes -> These should be plain classes which only transform the state they represent from the state of the integration to the state they represent.

A coordinator is a common way to hold state of a (part of the) integration. Those should exist in `coordinator.py`. Adding new entities or devices is not a responsibility of the entities (like the `ThinQSwitchEntity.create_entities` here) but a responsibility of the `async_setup_entry`s at the platform.

Reply:
Refactored.
Some code is moved into the PyPI library.